### PR TITLE
File Sharing using GridFS in MongoDB

### PR DESCRIPTION
**Pull request by girishgr8**
Added tutorial for the topic 

4 | MongoDB | Database | How to create a database on MongoDB Atlas and access if from your laptop?

### CODE REVIEW

**Comment by Kevinpaulose05** [comment]
Hi @girishgr8 . A microbyte is not a tutorial. It is Learn by doing the project. when a person takes a microbyte they need to work their way thru it, like they will have to follow a set of hints/guidelines to solve some mini-issues and also indirectly build something in the end that will facilitate the effective learning process. So we are sorry but we cannot be accepting this as this is totally against our guidelines as mentioned [before](https://docs.google.com/document/d/1WYVo-O1--RX3hGr6hxPRm6Mz_ky1ErNDUfUZv6oEw_s/edit#heading=h.o3gqdpr9r9ed).
Please refer to other microbytes(especially in the miscellaneous section) to better understand what I am implying.

**Comment by girishgr8** [comment]
Alright. I will make necessary changes considering the guidelines.

**Comment by amoghrajesh** [comment]
Hey @girishgr8!
Thanks for understanding.
Try to make the contributions asap as you already have the content, need a reordering basically.
Feel free to ping us on Slack in case of any queries 😄

**Comment by Kevinpaulose05** [comment]
@girishgr8 Pls work on my comments. We are eager to see ur completed contribution. U will get creator badges only if u do the needed latest by today midnight

**Comment by girishgr8** [comment]
> @girishgr8 Pls work on my comments. We are eager to see ur completed contribution. U will get creator badges only if u do the needed latest by today midnight

I will try to incorporate the changes, but I am quite confused about how to tweak in the changes to make my contribution not tutorialish.

**Comment by Kevinpaulose05** [comment]
> > @girishgr8 Pls work on my comments. We are eager to see ur completed contribution. U will get creator badges only if u do the needed latest by today midnight
> 
> I will try to incorporate the changes, but I am quite confused about how to tweak in the changes to make my contribution not tutorialish.

Thing is u need to revamp ur project as it is a tutorial. Like some application of MongoDB using its commands(in either shell or compass), like gridFS for file sharing, or image/pdf storing,MongoDB shell methods etc. But don't make it into a tutorial to complete that application. Make it Learning as well as hands-on doing

**Comment by girishgr8** [comment]
> > > @girishgr8 Pls work on my comments. We are eager to see ur completed contribution. U will get creator badges only if u do the needed latest by today midnight
> > 
> > 
> > I will try to incorporate the changes, but I am quite confused about how to tweak in the changes to make my contribution not tutorialish.
> 
> Thing is u need to revamp ur project as it is a tutorial. Like some application of MongoDB using its commands(in either shell or compass), like gridFS for file sharing, or image/pdf storing,MongoDB shell methods etc. But don't make it into a tutorial to complete that application. Make it Learning as well as hands-on doing

I have changed the tutorialish stuff and added hands-on learning for uploading files using gridfs & mongodb. Please guide me how to add the updated changes in this pull request.

**Comment by girishgr8** [comment]
# Introduction

### MongoDB

> MongoDB is a open-source document database and leading NoSQL database with the scalability and flexibility that you want with the querying and indexing.

### MongoDB Atlas

> MongoDB Atlas provides an easy way to host and manage your data in the cloud. MongoDB Atlas is a fully-managed cloud database.

---

> Ever wondered how to store large files like pdfs or images into databases ? If you have basic knowlegde of MySQL (A relational database), then you might know there is **blob** datatype which can be used to store pdfs. But how to do the same in MongoDB ?

> MySQL is relational database and hence it is not recommended to use for storing large amount of files, instead many professionals use MongoDB (A document-oriented NoSQL) database for it.

### Now, the solution for storing big files is using **GridFS with MongoDB**.

### **GridFS**

> GridFS is the MongoDB specification for storing and retrieving large files such as images, audio files, video files, etc. GridFS has the capability to store files even greater than its document size limit of 16MB.
> GridFS divides a file into chunks and stores each chunk(piece) of data in a separate document, each of maximum size 255K.

---

# Activities

### Activity 1: Setting up MongoDB cluster using MongoDB Atlas

[Follow the instructions in the document](https://docs.google.com/document/d/1R03KKVmihUU7nrjdTSZfwj6qLB-UvvR8OybtpjRWtFc/edit?usp=sharing)

---

### Activity 2: Import the libraries which will be required for our tasks:

-   This microbyte uses Python language for performing tasks.
-   Firstly open terminal and copy the following snippet & paste it in the terminal.

```bash
    $ pip install pymongo gridfs bson dnspython
```

1. Import MongoClient so that we can connect to a running mongod instance on our cluster.
2. Import GridFS from gridfs module so that we may put, get and delete files in MongoDB.
3. Import objectid from the bson library because we will be requiring it.
4. dnspython package is needed to make network calls to the cluster.

-   Open the command prompt and type **python** and click enter (This will turn your cmd into python shell):

```python
from pymongo import MongoClient
from gridfs import GridFS
from bson import objectid
import dns
```

---

### Activity 3: Connect to the cluster and access the database

> Now, we have imported the libraries, our task is to connect to the online cluster through our system.

**Task 1: Get the connection string to your cluster & connect to your database**

<details>
    <summary>Hint</summary>
While performing the Activity 1 above, you might have encountered with the connection string to your cluster and wondered about the use of it. This connection string is unique for every individual's cluster. Copy that connection string. Now use MongoClient to connect to it.

```python
connection_url = "mongodb+srv://CrioBytes:CrioBytes@criobytes.dyseu.mongodb.net/CrioBytes?retryWrites=true&w=majority"
conn = pymongo.MongoClient(connection_url)
# dbname will be present in your url (default db) named `test`
dbname = 'CrioBytes'
db = conn[dbname]
```

</details>
<br>

---

### Activity 4: Lets get started with interacting with our db & clear some basics

> **Task 1: Create the GridFS object**

-   **objects** ! The entire universe is made of objects !!!
-   Here, I refer object as creating the instance of GridFS class i.e. `GridFS()`.
-   `GridFS()` takes an argument which is the object (oops! again an object) of db which we created in Activity 3.

<details>
    <summary>Hint</summary>

```python
    gridfs = GridFS(db)
```

</details>

<br>

> **Task 2: Store simple string "Microbytes" in the database using gridfs**

-   Use `put()` method of gridfs instance we created to insert "Microbytes" in db
-   The `put()` method returns the object id of newly inserted object (may it be string or any file)
<details>
    <summary>Hint</summary>

```python
    obj = gridfs.put('MicroBytes')
```

</details>
<br>

**Note: This string is stored in `fs` namespace which is the default namespace used for storing data using GridFS**

> Curious about knowing what is **namespace** ?

-   A namespace is a system to have a unique name for each and every object.
-   At times, an object might be a variable or a method.
-   Lets break the keyword **namespace** as name (which means name, an unique identifier) + space(which tells us something related to scope). Here, a name might be of any method or variable and space depends upon the location from where is trying to access a variable or a method.
    <br>

---

### Activity 5: Diving deeper into GridFS

**Task 1: Create your own namespace**

-   Connect to cluster and create your own namespace while instantiating gridfs object
-   The `GridFS()` can also take another parameter which is the name of the **namespace** we wanna create.

<details>
<summary>Hint</summary>

```python
    fs = GridFS(db, "microbytes")
```

</details>

**Task 2: Insert any pdf file using gridfs**

-   Now, for inserting any file, we must have a reference pointing to that file.
-   In Python, we use `open()` function to open any file. This method returns a reference object to the file we opened.

```python
    f = open('/path_to_your_file/sample_bytes.pdf')
```

-   Use this reference object for inserting our pdf file.

<details>
<summary>Hint</summary>

```python
    file_id = fs.put(f, content_type='application/pdf', filename='sample_bytes.pdf')
    # you can provide the filename if you want to upload it with different name
```

Here, The put() method returned the file_id.

</details>
<br>

-   Now if you go to the database 'CrioBytes' in your cluster you fill find the collections named 'microbytes.files' & 'microbytes.chunks'

**Task 3: Read the uploaded file using the file_id object**

-   Can you use `get()` & `read()` method provided by `fs` object.
-   `get()` method returns a file-like-object, and to see the content, we can call `read()` method on this file-like-object.
-   Will `get()` method require some parameter associated with the file we wanna retrieve? Ofc yes ! To find the file we want, it takes in the file_id.

<details>
<summary>Hint</summary>

```python
    print(fs.get(file_id).read())
    # This will print the content of the file which has id as file_id
```

</details>
<br>

**Task 4: Let's insert txt file into MongoDB**

-   Since you have worked with inserting pdf files, this task shall not be difficult one.
-   You will perform something just like below :

```python
    f = open('/path_to_your_file/sample_bytes.pdf')
    fs.put(f, filename="sample_text.txt")
```

> What if we want to insert the file which is not present on the local system ? Can we insert some text file by inputting the data into that text file on the fly ? (dynamically) !! This might be necessary in some crucial tasks !!! So, lets see how to do so !

-   You might remember, we inserted the simple string "MicroBytes" at the start, and in subsequent task, we inserted text file. Now, the solution is simple, combine/tweak both ?

```python
    fs.put("Hello, This is a sample string", filename="new.txt")
```

<br>

**Task 5: Find a file in our database & read its contents**

-   As mentioned Task 2 of Activity 5, the files are stored in db.namespace.files collection. Use this collection to find the file stored.
-   Use `find_one()` method to find the document inside the collection.
-   Use `get()` & `read()` to read the contents of file.
<details>
    <summary>Hint</summary>

```python
    file_id = db.microbytes.files.find_one({"filename" : "sample_text.txt"},{"_id":1})
    print(fs.get(file_id['_id']).read())
    # Contents of the file `sample_text.txt` will be printed
```

</details>
<br>

**Task 6: Delete the file**

-   Now, suppose we want to delete the stored file, then use `delete()` method of `fs` object.
-   `delete()` method will also require the file_id of the file one wants to delete (Remember ! The `get()` method also needed file_id)
<details>
    <summary>Hint</summary>

```python
    fs.delete(file_id)
```

</details>
<br>

---

Hooray 🎉🎉 We have successfully setup clustes, created database,uploaded, retrieved, deleted files using GridFS file system it through our personal laptop/desktop.

# Summary

We have created a database on MongoDB Atlas and accessed it from your own laptop.

# References

1. https://docs.atlas.mongodb.com/getting-started/
2. https://www.w3schools.com/python/python_mongodb_getstarted.asp
3. https://www.tutorialspoint.com/mongodb/mongodb_gridfs.htm
4. https://psabhay.com/posts/mongodb/mongodb-gridfs-using-python/

**Comment by Kevinpaulose05** [comment]
Do u use command line or giuhub directly?
 
Also there are 2 bad practices here:
1. U passed ur code as a comment. In other scenarios someone may copy it without ur consent; and also it is not plesaing for someone to follow the comments this way
2. Always make a pull request from a branch. That way ur main repo is safe

Keep these in mind.

**Comment by girishgr8** [comment]
> Do u use command line or giuhub directly?
> 
> Also there are 2 bad practices here:
> 
> 1. U passed ur code as a comment. In other scenarios someone may copy it without ur consent; and also it is not plesaing for someone to follow the comments this way
> 2. Always make a pull request from a branch. That way ur main repo is safe
> 
> Keep these in mind.

I use Github directly. Sorry for pasting in the code directly, I was unaware that such thing is a bad practice. I will make a pull request again with updated code. Is that okay ?

**Comment by Kevinpaulose05** [comment]
Delete one file/folder and add its subsequent file(s). Never leave the repo empty

**Comment by Kevinpaulose05** [comment]
opening a new PR will make our entire conversation go futile for other ppl to learn from

**Comment by girishgr8** [comment]
I have updated the repo with the changes and added the file in my forked repo. Will you please review the same?

**Comment by Kevinpaulose05** [comment]
Don't use the light shading(>) markdown everywhere. Us it for special mentions/notes only. Re-format ur MB pls

**Comment by Kevinpaulose05** [comment]
Give some article link for relational/noSQL database in MongoDB atlas section

**Comment by Kevinpaulose05** [comment]
Give a section of pre-requisites. MongoDB basic/ database basic, basic python, etc. Also mention python3 shud be installed

**Comment by Kevinpaulose05** [comment]
Post a question regarding how big a BSON file can be uploaded. Does gridfs support multiple document transactions? Address how it stores data(in chunks), give some illustrative explanation. Give some other trivia that u find

File: MongoDB/src/README.md
Comment (Kevinpaulose05) [inline MongoDB/src/README.md]: Type- clustes shud be 'clusters'

**Comment by Kevinpaulose05** [comment]
**_NOTE:_** Always use git(i.e. from cmd line or git bash). It is very powerful and will help u in managing ur code around github better. Do learn it

**Comment by Kevinpaulose05** [comment]
Make the aforementioned changes. rest looks fine for now :p

File: MongoDB/src/README.md
Comment (girishgr8) [reply MongoDB/src/README.md]: > Make the aforementioned changes. rest looks fine for now :p

Ya I will update the code with required changes and remove typos & structure the content properly

**Comment by girishgr8** [comment]
Updates done:

1. Pre-requisites added
2. Removed unnecessary light shaded text
3. Added article link for basic difference between relational & nosql db
4. BSON-document max size limit
5. Question & reasoning on why GridFS doesnot support multi-doc transactions
6. Basic explanation on how files are stored in chunks.
7. Fixed the typos


@Kevinpaulose05  
Please review the changes.

**Comment by girishgr8** [comment]
Are any changes required ?

**Comment by Kevinpaulose05** [comment]
The first place u mentioned BSON, give a link to it. Newbies may think it as JSON maybe ;)
in pre requisites give a python installation link

**Comment by girishgr8** [comment]
> The first place u mentioned BSON, give a link to it. Newbies may think it as JSON maybe ;)
> in pre requisites give a python installation link

Yaaa alright.... I will add that !

**Comment by Kevinpaulose05** [comment]
Can u come up with some challenege based on gridfs? So that learners can solve it. Try to think of something. It wud be good.
Also kindly centre align both images(chunk one and immediate next one-shards one)

**Comment by Kevinpaulose05** [comment]
Also just add some vibrant images like file sharing ginto database, or any funny photo that will make ur microbyte an eye-catcher? Just a suggestion. Work on the previous ones first :)

**Comment by girishgr8** [comment]
> Can u come up with some challenege based on gridfs? So that learners can solve it. Try to think of something. It wud be good.
> Also kindly centre align both images(chunk one and immediate next one-shards one)

Ya actually there can be some challenge related to adding the same file to the database using gridfs again and again and asking user some task related to it ?

**Comment by Kevinpaulose05** [comment]
Take time if needed. There is no hurry. I have added the label. So ur PR I counted in the official contest(revert back here if not). We will accept submissions as and when u like. But kindly don't stretch it too much.

**Comment by Kevinpaulose05** [comment]
> > Can u come up with some challenege based on gridfs? So that learners can solve it. Try to think of something. It wud be good.
> > Also kindly centre align both images(chunk one and immediate next one-shards one)
> 
> Ya actually there can be some challenge related to adding the same file to the database using gridfs again and again and asking user some task related to it ?

Yup. that will do

**Comment by girishgr8** [comment]
> Take time if needed. There is no hurry

My contribution won't count in this repository if it exceeds time ?

**Comment by girishgr8** [comment]
> Take time if needed. There is no hurry. I have added the label. So ur PR I counted in the official contest(revert back here if not). We will accept submissions as and when u like. But kindly don't stretch it too much.

Sure. I will try to update it by today itself.

**Comment by girishgr8** [comment]
@Kevinpaulose05 
Please review the changes which were suggested by you.

**Comment by Kevinpaulose05** [comment]
Can u change ur MB name to something like file sharnig with mongodb or something. MongoDb is a generic name

**Comment by girishgr8** [comment]
> Can u change ur MB name to something like file sharnig with mongodb or something. MongoDb is a generic name

Sure ! I will change it **File Sharing using GridFS & MongoDB** ? Is that okay ?

**Review by Kevinpaulose05 — APPROVED** [review]
awesome work @girishgr8 . Two things to do **after merging**: 
1. Open issues related to ur MB
2. Also, kindly showcase your MicroByte in the **#microbyte-showcase** channel over at our slack workspace, using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md) template.

Also, kindly **fill** [this](https://forms.gle/mE9sB7xR577e2QmG7) form.

**Comment by Kevinpaulose05** [comment]
> > Can u change ur MB name to something like file sharnig with mongodb or something. MongoDb is a generic name
> 
> Sure ! I will change it **File Sharing using GridFS & MongoDB** ? Is that okay ?

gridfs in MongoDB? Ig this is right?

**Comment by girishgr8** [comment]
> > > Can u change ur MB name to something like file sharnig with mongodb or something. MongoDb is a generic name
> > 
> > 
> > Sure ! I will change it **File Sharing using GridFS & MongoDB** ? Is that okay ?
> 
> gridfs in MongoDB? Ig this is right?

Ya I did the changes .. Is this name okay ?

**Comment by girishgr8** [comment]
> awesome work @girishgr8 . Two things to do **after merging**:
> 
> 1. Open issues related to ur MB
> 2. Also, kindly showcase your MicroByte in the **#microbyte-showcase** channel over at our slack workspace, using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md) template.
> 
> Also, kindly **fill** [this](https://forms.gle/mE9sB7xR577e2QmG7) form.

Thank you for guiding me so nicely and suggesting the  good changes.

**Review by archithdwij — APPROVED** [review]
Awesome work @girishgr8 . I love your approach. 💯

**Comment by girishgr8** [comment]
> Awesome work @girishgr8 . I love your approach.

Thank you so much !

**Review by YourAKShaw — CHANGES_REQUESTED** [review]
@girishgr8 The content looks pretty decent now. I really like the way you've laid out the entire content of yours. Nice value added to the learning scope. :100: 

Just one minor change though. Kindly rename your folder to specifically state the purpose of the MicroByte in a very brief manner. The current title `MongoDB` isn't good enough.

Just make this change, and we'll get the PR merged.

**Comment by girishgr8** [comment]
> @girishgr8 The content looks pretty decent now. I really like the way you've laid out the entire content of yours. Nice value added to the learning scope. 💯
> 
> Just one minor change though. Kindly rename your folder to specifically state the purpose of the MicroByte in a very brief manner. The current title `MongoDB` isn't good enough.
> 
> Just make this change, and we'll get the PR merged.

I renamed the microbyte folder to **"File Sharing using GridFS in MongoDB"** is that okay ?

**Comment by YourAKShaw** [comment]
Yeah, cool.

**Review by YourAKShaw — APPROVED** [review]

**Comment by YourAKShaw** [comment]
@all-contributors kindly add @girishgr8 for:

- [x] code
- [x] content
- [x] ideas
- [x] doc

**Comment by girishgr8** [comment]
Thank you @Kevinpaulose05 @amoghrajesh @Ak-Shaw @archithdwij  for guiding & helping me out in creating such good microbyte. It was a wonderful experience !!

**Comment by YourAKShaw** [comment]
The feeling is mutual.

**Comment by YourAKShaw** [comment]
> awesome work @girishgr8 . Two things to do **after merging**:
> 
>     1. Open issues related to ur MB
> 
>     2. Also, kindly showcase your MicroByte in the **#microbyte-showcase** channel over at our slack workspace, using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md) template.
> 
> 
> Also, kindly **fill** [this](https://forms.gle/mE9sB7xR577e2QmG7) form.

Kindly do these.

**Comment by girishgr8** [comment]
> > awesome work @girishgr8 . Two things to do **after merging**:
> > ```
> > 1. Open issues related to ur MB
> > 
> > 2. Also, kindly showcase your MicroByte in the **#microbyte-showcase** channel over at our slack workspace, using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md) template.
> > ```
> > 
> > 
> > Also, kindly **fill** [this](https://forms.gle/mE9sB7xR577e2QmG7) form.
> 
> Kindly do these.

The slack workspace link is broken. The page is not found ! Also, just a small doubt, I should add further improvement scope for this MB in the issues section right ?

**Comment by YourAKShaw** [comment]
Try this: https://join.slack.com/t/criodo-world/shared_invite/zt-ijwn6acw-oF7qy9eiYGgJJQPh5QuPGQ

**Comment by YourAKShaw** [comment]
Yeah, kindly try to create improvement issues if you have ideas.

**Pull request merged**
